cmd/package_managers.go

```diff
@@ -138,3 +205,145 @@ func fetchGitRepositoryFromRubyGems(packageName string, manager packageManagerCl
+	if len(strings.TrimSpace(packageSpecResults.Metadata.ProjectURL)) != 0 {
+		projectURL = packageSpecResults.Metadata.ProjectURL
+	}
+	if isSupportedProjectURL(projectURL) {
+		return strings.TrimSuffix(projectURL, ".git"), nil
+	}
+	return "", sce.WithMessage(sce.ErrScorecardInternal,
+		fmt.Sprintf("source repo is not defined for nuget package %v", packageName))
+}
+
+func isSupportedProjectURL(projectURL string) bool {
+	pattern := `^(?:https?://)?(?:www\.)?(?:github|gitlab)\.com/([A-Za-z0-9_\.-]+)/([A-Za-z0-9_\./-]+)$`
+	regex := regexp.MustCompile(pattern)
+	return regex.MatchString(projectURL)
+}
+
+// Gets the latest listed nuget version of a package, based on the protocol defined at
+// https://learn.microsoft.com/en-us/nuget/api/package-base-address-resource#enumerate-package-versions
+func getLatestListedVersion(baseURL, packageName string, manager packageManagerClient) (string, error) {
+	resPackageRegistrationIndex, err := manager.Get(baseURL+"%s/index.json", packageName)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to get nuget package registration index json: %v", err))
+	}
+	defer resPackageRegistrationIndex.Body.Close()
+	packageRegistrationCatalogRoot := &nugetPackageRegistrationCatalogRoot{}
+	err = json.NewDecoder(resPackageRegistrationIndex.Body).Decode(packageRegistrationCatalogRoot)
+	if err != nil {
+		return "", sce.WithMessage(sce.ErrScorecardInternal,
+			fmt.Sprintf("failed to parse package registration index json: %v", err))
+	}
+	return getLatestListedVersionFromPackageRegistrationPages(packageRegistrationCatalogRoot.Pages, manager)
+}
+
+func getLatestListedVersionFromPackageRegistrationPages(pages []nugetPackageRegistrationCatalogPage,
+	manager packageManagerClient,
+) (string, error) {
+	for pageIndex := len(pages) - 1; pageIndex >= 0; pageIndex-- {
+		page := pages[pageIndex]
+		if page.Packages == nil {
+			respPage, err := manager.GetURI(page.ID)
+			if err != nil {
+				return "", sce.WithMessage(sce.ErrScorecardInternal,
+					fmt.Sprintf("failed to get nuget package registration page: %v", err))
+			}
+			defer respPage.Body.Close()
+			err = json.NewDecoder(respPage.Body).Decode(&page)
+
+			if err != nil {
+				return "", sce.WithMessage(sce.ErrScorecardInternal,
+					fmt.Sprintf("failed to parse nuget package registration page: %v", err))
+			}
+		}
+		for packageIndex := len(page.Packages) - 1; packageIndex >= 0; packageIndex-- {
+			semVerString, preRelease := getNugetSemVer(page.Packages[packageIndex].Entry.Version)
+			// skipping non listed and pre-releases
+			if page.Packages[packageIndex].Entry.Listed && len(strings.TrimSpace(preRelease)) == 0 {
+				return semVerString, nil
+			}
+		}
+	}
+	return "", sce.WithMessage(sce.ErrScorecardInternal,
+		"failed to get a listed version for package")
+}
+
+// Nuget semver diverges from Semantic Versioning.
+// This method returns the Nuget represntation of version and pre release strings.
+// nolint
```

Comment:
can you be specific about which linter you're ignoring and why?

```
//nolint:lintername // some comment why
```

Reply:
done